include/pybind11/pybind11.h

```diff
@@ -1954,30 +1954,48 @@ inline std::pair<decltype(internals::registered_types_py)::iterator, bool> all_t
     return res;
 }
 
-template <typename Iterator, typename Sentinel, bool KeyIterator, return_value_policy Policy>
+template <typename Access, return_value_policy Policy, typename Iterator, typename Sentinel, typename ValueType, typename... Extra>
 struct iterator_state {
     Iterator it;
     Sentinel end;
     bool first_or_done;
 };
 
-PYBIND11_NAMESPACE_END(detail)
+template <typename Iterator>
+struct iterator_access {
+    detail::remove_cv_t<decltype(*std::declval<Iterator>())> operator()(const Iterator it) {
+        return *it;
+    }
+};
 
-/// Makes a python iterator from a first and past-the-end C++ InputIterator.
-template <return_value_policy Policy = return_value_policy::reference_internal,
+template <typename Iterator>
+struct iterator_key_access {
+    detail::remove_cv_t<decltype((*std::declval<Iterator>()).first)> operator()(const Iterator it) {
+        return (*it).first;
+    }
+};
+
+template <typename Iterator>
+struct iterator_value_access {
+    detail::remove_cv_t<decltype((*std::declval<Iterator>()).second)> operator()(const Iterator it) {
+        return (*it).second;
+    }
+};
+
+template <typename Access,
+          return_value_policy Policy,
           typename Iterator,
           typename Sentinel,
-#ifndef DOXYGEN_SHOULD_SKIP_THIS  // Issue in breathe 4.26.1
-          typename ValueType = decltype(*std::declval<Iterator>()),
-#endif
+          typename ValueType,
           typename... Extra>
-iterator make_iterator(Iterator first, Sentinel last, Extra &&... extra) {
-    using state = detail::iterator_state<Iterator, Sentinel, false, Policy>;
+iterator make_iterator_impl(Iterator first, Sentinel last, Extra &&... extra) {
+    using state = detail::iterator_state<Access, Policy, Iterator, Sentinel, ValueType, Extra...>;
+    // TODO: state captures only the types of Extra, not the values
 
     if (!detail::get_type_info(typeid(state), false)) {
         class_<state>(handle(), "iterator", pybind11::module_local())
             .def("__iter__", [](state &s) -> state& { return s; })
-            .def("__next__", [](state &s) -> ValueType {
+            .def("__next__", [](state &s) -> detail::remove_cv_t<ValueType> {
```

Comment:
@rwgk I think we need to make sure that this doesnt' cause those edge case issues we were having with removing the const in return types for Eigen

Reply:
I could run this through our global testing tonight (after I got a chance to take a closer look at the code).

Reply:
@Skylion007 could you elaborate on what those problems were? Is it something that might only show up when iterating over a container of Eigen objects?

Reply:
Long story short, Eigen has a special caster that makes it so const Eigen objects are to Numpy arrays with the writable flag off, while non-const Eigen objects have them on so this can actually break some tests even though it's against C++ guidelines / behavior.

Reply:
So what would a test for that look like? Something like binding `std::vector<const Eigen::Vector3f>`, passing it to Python and checking that the writeable flag is off when iterating it? For that matter I'm not even sure it's valid to have a const value type in an STL container.

Reply:
To be on the safe side I've removed all the `remove_cv_t` stuff and put in NOLINTNEXTLINE instead. I also discovered that subtleties in `decltype` meant in master, `make_key_iterator` was making copies rather than references. I think that should be fixed in this PR (it's more important for `make_value_iterator`, since map values are mutable while keys are not), but I still need to add some tests.

Reply:
Shall we add a unit test asserting the key is not copied (e.g. delete the copy constructor)?

Reply:
> Shall we add a unit test asserting the key is not copied (e.g. delete the copy constructor)?

That's exactly the test I've added. I just need to see why it's failing on some compilers in the CI matrix but not others.